Make JavaScriptEvaluationResult platform independent

achristensen07 opened the pull request:
#### 499215d196f3a1d8ff913989f1ccb427b5c790ee
<pre>
Make JavaScriptEvaluationResult platform independent
<a href="https://bugs.webkit.org/show_bug.cgi?id=295744">https://bugs.webkit.org/show_bug.cgi?id=295744</a>
<a href="https://rdar.apple.com/155561490">rdar://155561490</a>

Reviewed by Brady Eidson.

In the initial development of multi-process WebKit, we used SerializedScriptValue to send values
from JavaScript in the web content process to the UI process, which then needed to make a JSContext
to deserialize the SerializedScriptValue back to a JSValue then extract the data from it.
This was cumbersome and problematic for many reasons.

With the introduction of WKWebView, we abstracted this process away from API clients, and we give
them one of a number of types of objects in the same API language instead.  In the case of WKWebView&apos;s
evaluateJavaScript, we provide one of these types: NSNumber, NSString, NSDate, NSArray, NSDictionary,
NSNull, and nil.

With 281812@main I migrated the legacy C API to this new paradigm as well.  With 281960@main I
removed the concept of SerializedScriptValue from the API surface.  I&apos;ve also introduced
JavaScriptEvaluationResult to give a simpler IPC format that is still powerful enough to express
all supported JSValue return types that we transport to the UI process, including things like
arrays and dictionaries that contain themselves or contain objects that form graph loops.
JavaScript is a dynamic language, and we supported those things before, so we continue to support them.

With 297201@main I removed the ObjC-specific pieces of JavaScriptEvaluationResult to make it able to
be used on platforms without ObjC support.

This PR completes the journey by making JavaScriptEvaluationResult used the same on all platforms.

Many thanks to Carlos Garcia Campos who helped fix some glib test failures.

* Source/WebKit/Shared/API/APISerializedScriptValue.cpp:
(API::SerializedScriptValue::deserializationContext):
* Source/WebKit/Shared/API/APISerializedScriptValue.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::toAPI):
(WebKit::JavaScriptEvaluationResult::toWK):
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::toVariant):
(WebKit::JavaScriptEvaluationResult::toJS):
(WebKit::JavaScriptEvaluationResult::legacySerializedScriptValue const): Deleted.
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
(WebKit::JavaScriptEvaluationResult::map const):
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult): Deleted.
(WebKit::JavaScriptEvaluationResult::wireBytes const): Deleted.
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in:
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRunJavaScriptWithParams):
(parseAsyncFunctionArguments):
* Source/WebKit/UIProcess/API/gtk/WebKitRemoteInspectorProtocolHandler.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):

Canonical link: <a href="https://commits.webkit.org/297498@main">https://commits.webkit.org/297498@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71ab657fe973e003154997f17cf79bc8f9121d38

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111993 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31663 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22149 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118016 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/62229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32340 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40241 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/85076 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/62229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114940 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25830 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100773 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65508 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/25146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61865 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95215 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18990 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/121332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39026 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29035 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/121332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39407 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97032 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/121332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38949 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/16737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18051 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38920 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44432 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/38557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41885 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/40273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
```diff
@@ -206,9 +214,9 @@ JavaScriptEvaluationResult::JavaScriptEvaluationResult(JSGlobalContextRef contex
+JSValueRef JavaScriptEvaluationResult::toJS(JSGlobalContextRef context, Value&& root)
 {
-    return WTF::switchOn(WTFMove(root), [&] (NullType) -> JSValueRef {
+    return WTF::switchOn(WTFMove(root), [&] (EmptyType) -> JSValueRef {
```
This is causing most of the failures in glib api tests. The empty type is ignored and Null value is always returned. Creating an undefined value when EmptyType is Undefined fixes some of the failures. I'm still investigating the others.

carlosgcampos commented on Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
```diff
@@ -264,41 +272,91 @@ Protected<JSValueRef> JavaScriptEvaluationResult::toJS(JSGlobalContextRef contex
-    return JavaScriptEvaluationResult { context, value };
+    if (!variant)
+        return std::nullopt;
+    return JavaScriptEvaluationResult(variant);
```
The other problem is here, we need to check first if the variant can be serialized like cocoa does. I'll attach a patch with both fixes.

achristensen07 committed:
Make JavaScriptEvaluationResult platform independent
https://bugs.webkit.org/show_bug.cgi?id=295744
rdar://155561490

Reviewed by Brady Eidson.

In the initial development of multi-process WebKit, we used SerializedScriptValue to send values
from JavaScript in the web content process to the UI process, which then needed to make a JSContext
to deserialize the SerializedScriptValue back to a JSValue then extract the data from it.
This was cumbersome and problematic for many reasons.

With the introduction of WKWebView, we abstracted this process away from API clients, and we give
them one of a number of types of objects in the same API language instead.  In the case of WKWebView's
evaluateJavaScript, we provide one of these types: NSNumber, NSString, NSDate, NSArray, NSDictionary,
NSNull, and nil.

With 281812@main I migrated the legacy C API to this new paradigm as well.  With 281960@main I
removed the concept of SerializedScriptValue from the API surface.  I've also introduced
JavaScriptEvaluationResult to give a simpler IPC format that is still powerful enough to express
all supported JSValue return types that we transport to the UI process, including things like
arrays and dictionaries that contain themselves or contain objects that form graph loops.
JavaScript is a dynamic language, and we supported those things before, so we continue to support them.

With 297201@main I removed the ObjC-specific pieces of JavaScriptEvaluationResult to make it able to
be used on platforms without ObjC support.

This PR completes the journey by making JavaScriptEvaluationResult used the same on all platforms.

Many thanks to Carlos Garcia Campos who helped fix some glib test failures.

* Source/WebKit/Shared/API/APISerializedScriptValue.cpp:
(API::SerializedScriptValue::deserializationContext):
* Source/WebKit/Shared/API/APISerializedScriptValue.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::toAPI):
(WebKit::JavaScriptEvaluationResult::toWK):
(WebKit::JavaScriptEvaluationResult::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::toVariant):
(WebKit::JavaScriptEvaluationResult::toJS):
(WebKit::JavaScriptEvaluationResult::legacySerializedScriptValue const): Deleted.
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
(WebKit::JavaScriptEvaluationResult::map const):
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult): Deleted.
(WebKit::JavaScriptEvaluationResult::wireBytes const): Deleted.
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in:
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewRunJavaScriptWithParams):
(parseAsyncFunctionArguments):
* Source/WebKit/UIProcess/API/gtk/WebKitRemoteInspectorProtocolHandler.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScript):

Canonical link: https://commits.webkit.org/297498@main

mcrha commented on Source/WebKit/Shared/glib/JavaScriptEvaluationResultGLib.cpp:
```diff
@@ -0,0 +1,137 @@
+        GVariant* value;
+        while (g_variant_iter_loop(&iter, "{&sv}", &key, &value)) {
+            if (!key || !isSerializable(value))
+                return false;
```
this leaks `value`, according to the `g_variant_iter_loop` documentation

mcatanzaro replied:
Thanks, created https://github.com/WebKit/WebKit/pull/50029
